lib/reek/smells/smell_detector.rb

```diff
@@ -120,20 +76,45 @@ def config_for(ctx)
         ctx.config_for(self.class)
       end
 
-      # :reek:FeatureEnvy
       def smell_warning(options = {})
-        context = options.fetch(:context)
-        exp = context.exp
-        ctx_source = exp.loc.expression.source_buffer.name
-        SmellWarning.new(self,
-                         source: ctx_source,
-                         context: context.full_name,
-                         lines: options.fetch(:lines),
-                         message: options.fetch(:message),
-                         parameters: options.fetch(:parameters, {}))
+        SmellWarning.build self, options
```

Comment:
I don't think I agree. The contents of options is specific to the context of the smell detector.

With your change, we have two constructors: SmellWarning.build and SmellWarning.new, with similar but different parameters. I put this logic specifically outside of SmellWarning so that class doesn't need to know about the internals of Context.